### PR TITLE
memory tracking (internal self-review)

### PR DESCRIPTION
**Pull request by tjjordan**
## Intro
This PR introduces new logic to track the temporary memory usage during query execution. At a high level, we introduce a new meta table that displays temporary memory usage data as a series of allocations and de-allocations over time.

We track allocations and deallocations at the level of operators and data structures, as shown in the made up example below:

operator_type | operator_data_structure | timestamp |  amount
--- | --- | --- | ---
JoinHash | input_materialization_radix_container | 1658016000011568765 | 10
JoinHash | input_materialization_radix_container | 1658016000011568765 | -10

Currently, memory tracking is only implemented for the hash join. We discuss later how memory tracking can be implemented for additional operators.

The memory tracking is deactivated by default and can be activated explicitly for the benchmarks by adding the command line flag `--memory-tracking`. The tracking data is not yet stored as part of the benchmark output, this is an open TODO.

## Main Changes
We introduce a new class `TrackingMemoryResource` that is derived from `boost::container::pmr::memory_resource` and can be used in `PolymorphicAllocator`s. `TrackingMemoryResource` instances track how much memory (bytes) was allocated or deallocated at which point in time. Timestamps are recorded using the `std::chrono::system_clock` (just like for the logging, cf. `LogManager`).

We introduce a new manager: the `MemoryResourceManager`. Operators can request `TrackingMemoryResource`s for their data structures from a central `MemoryResourceManager` instance that lives in the `Hyrise` singleton. The `MemoryResourceManager` creates new `TrackingMemoryResource`s, vends them to operators and stores them in a `tbb::concurrent_vector`.  We use a `tbb::concurrent_vector` so that concurrently running operators can interact with the memory resource manager without running into issues.

The new meta table `MetaTemporaryMemoryUsageTable` obtains `TrackingMemoryResource`s from the `MemoryResourceManager`. It then collects the recorded allocations from the `TrackingMemoryResource`s and builds the meta table. The timestamps are displayed as longs in unix time at nanosecond granularity (just like for the `MetaLogTable`).

We used the implementations provided by the boost library because memory resources aren't yet included in libc++.

In addition to this, we exchanged the hash map used by the `JoinHash` operator.
There was no way to pass a custom constructor to the previously used `bytell_hash_map`, which is why tracking its memory usage would have been impossible.
Since we are not the first to use `PolymorphicAllocator`s in Hyrise we decided to exchange the hash map.
After doing some benchmarks with several hash maps, Tessil's `robin_map` seemed to be a good option.
Except the removal of a `shrink_to_fit()` call which is not supported by Tessil's hash maps, no other changes to the `JoinHash` code were necessary.
The hash map exchange may also bring slight performance increases, but it is necessary to reproduce them on reliable benchmark hardware.

To avoid circular dependencies we moved the `OperatorType` enum to types.hpp. This also made sense since there were already some enumerations defined.

## How to Implement Memory Tracking in an operator
We recommend creating an `alloc` method for each operator that we want to support memory tracking for. The alloc method should take the data structure name as a parameter and return a polymorphic allocator. The operator type of the underlying memory resource should be set to the name of the operator. The data structure name parameter can have a default value to simplify the call sites when detailed tracking is not needed. 

As an example, here is the alloc method for the hash join:
```c++
template <typename T>
PolymorphicAllocator<T> alloc(const std::string& operator_data_structure) {
  return PolymorphicAllocator<T>{
      Hyrise::get().memory_resource_manager.get_memory_resource(OperatorType::JoinHash, operator_data_structure)};
}
```
The alloc method can be used as follows:

```c++
auto my_mem_tracking_vector = pmr_vector<int>(alloc<int>("my_data_structure_name"));
```

## How to Try
To see the temporary memory tracking in action, you just need to perform a hash join.
E.g. a join on some data generated for the TPCH benchmark using the Hyrise console:
```
./hyriseConsole
generate_tpch 0.1;
select * from customer join orders on c_custkey=o_custkey limit 10;
```
After this, the temporary memory usage is stored in the meta memory table.
This data can be queried, e.g. with:
```
select * from meta_temporary_memory_usage;
select operator_type, operator_data_structure, sum(amount) from meta_temporary_memory_usage group by operator_type, operator_data_structure;
```

Please note: To enable playing around we enabled the temporary memory tracking by default.
Later, it should be deactivated by default. Users can then activate it manually in the console or via user callable functions in the server.

## Open TODOs
- #31
- #32
- #14
- #33

### CODE REVIEW

File: src/lib/utils/meta_tables/meta_temporary_memory_usage_table.cpp [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+MetaTemporaryMemoryUsageTable::MetaTemporaryMemoryUsageTable()
+    : AbstractMetaTable(TableColumnDefinitions{{"operator_type", DataType::String, false},
+                                               {"operator_data_structure", DataType::String, false},
+                                               {"timestamp", DataType::Long, false},
+                                               {"amount", DataType::Long, false}}) {}
+
+const std::string& MetaTemporaryMemoryUsageTable::name() const {
+  static const auto name = std::string{"memory"};
+  return name;
+}
```
Comment (twollnik) [inline src/lib/utils/meta_tables/meta_temporary_memory_usage_table.cpp:18]: @tjjordan I wonder if we should rename the table to `"temporary_memory_usage"` to be consistent with the updated class name `MetaTemporaryMemoryUsageTable`.

File: src/test/lib/operators/join_hash/join_hash_steps_test.cpp [hunk 2]
```diff
@@ -282,11 +282,11 @@ TEST_F(JoinHashStepsTest, RadixClusteringOfNulls) {
 }
 
 TEST_F(JoinHashStepsTest, BuildRespectsBloomFilter) {
-  std::vector<std::vector<size_t>> histograms;  // Ignored in this test
-  BloomFilter output_bloom_filter;              // Ignored in this test
+  pmr_vector<pmr_vector<size_t>> histograms;              // Ignored in this test
+  pmr_dynamic_bitset<unsigned long> output_bloom_filter;  // Ignored in this test
```
Comment (twollnik) [inline src/test/lib/operators/join_hash/join_hash_steps_test.cpp]:
```suggestion
  BloomFilter output_bloom_filter;  // Ignored in this test
```

File: src/test/lib/operators/join_hash/join_hash_steps_test.cpp [hunk 3]
```diff
@@ -282,11 +282,11 @@ TEST_F(JoinHashStepsTest, RadixClusteringOfNulls) {
...
 
   // Fill input_bloom_filter
-  BloomFilter input_bloom_filter(BLOOM_FILTER_SIZE);
+  pmr_dynamic_bitset<unsigned long> input_bloom_filter(BLOOM_FILTER_SIZE);
```
Comment (twollnik) [inline src/test/lib/operators/join_hash/join_hash_steps_test.cpp]:
```suggestion
  BloomFilter input_bloom_filter(BLOOM_FILTER_SIZE);
```

File: src/test/lib/operators/join_hash/join_hash_steps_test.cpp [hunk 4]
```diff
@@ -313,8 +313,8 @@ TEST_F(JoinHashStepsTest, ThrowWhenNoNullValuesArePassed) {
   }
 
   size_t radix_bit_count = 0;
-  std::vector<std::vector<size_t>> histograms;
-  BloomFilter bloom_filter;  // Ignored in this test
+  pmr_vector<pmr_vector<size_t>> histograms;
+  pmr_dynamic_bitset<unsigned long> bloom_filter;  // Ignored in this test
```
Comment (twollnik) [inline src/test/lib/operators/join_hash/join_hash_steps_test.cpp]:
```suggestion
  BloomFilter bloom_filter;  // Ignored in this test
```

File: src/test/lib/operators/join_hash/join_hash_types_test.cpp [hunk 5]
```diff
@@ -24,7 +24,7 @@ void test_hash_map(const std::vector<T>& values) {
 
   // Build a BloomFilter that cannot be used to skip any entries by creating a BloomFilter with every value being false
   // and using bitwise negation (~x).
-  auto bloom_filter = ~BloomFilter(BLOOM_FILTER_SIZE);
+  auto bloom_filter = ~pmr_dynamic_bitset<unsigned long>(BLOOM_FILTER_SIZE);
```
Comment (twollnik) [inline src/test/lib/operators/join_hash/join_hash_types_test.cpp]:
```suggestion
  auto bloom_filter = ~BloomFilter(BLOOM_FILTER_SIZE);
```

File: src/test/lib/sql/sql_pipeline_statement_test.cpp [hunk 6]
```diff
@@ -516,7 +516,6 @@ TEST_F(SQLPipelineStatementTest, GetResultTableJoin) {
 TEST_F(SQLPipelineStatementTest, GetResultTableWithScheduler) {
   auto sql_pipeline = SQLPipelineBuilder{_join_query}.create_pipeline();
   auto statement = get_sql_pipeline_statements(sql_pipeline).at(0);
-
```
Comment (twollnik) [inline src/test/lib/sql/sql_pipeline_statement_test.cpp:519]: let's leave this file unchanged to keep it out of the PR

File: src/test/lib/utils/memory_resource_manager_test.cpp [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+  const auto memory_resource_ptr_2 =
+      memory_resource_manager.get_memory_resource(OperatorType::Mock, "my_data_structure");
+  ASSERT_EQ(memory_resource_manager.memory_resources().size(), 2);
+
+  // we expect
+  ASSERT_NE(memory_resource_ptr_1, memory_resource_ptr_2);
+}
+
+TEST_F(MemoryResourceManagerTest, ConcurrentCallsAreHandledCorrectly) {
+  // get a memory resource, perform an allocation and a deallocation. Will be executed by multiple threads simultaneously.
```
Comment (twollnik) [inline src/test/lib/utils/memory_resource_manager_test.cpp]:
```suggestion
  // Get a memory resource, perform an allocation and a deallocation. Will be executed by multiple threads simultaneously.
```

File: src/test/lib/utils/memory_resource_manager_test.cpp [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+  auto fetch_and_use_a_memory_resource = [&](const uint8_t indx) {
+    auto memory_resource =
+        memory_resource_manager.get_memory_resource(OperatorType::Mock, "my_data_structure_" + std::to_string(indx));
+    auto mem_ptr = memory_resource->allocate(indx + 1);
+    memory_resource->deallocate(mem_ptr, indx + 1);
+  };
+
+  // Create a few threads that interact with the memory resource manager and obtained memory resources
```
Comment (twollnik) [inline src/test/lib/utils/memory_resource_manager_test.cpp:58]:
```suggestion
  // Create a few threads that interact with the memory resource manager and obtained memory resources.
```

File: src/test/lib/memory/tracking_memory_resource_test.cpp [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+namespace opossum {
+
+class TrackingMemoryResourceTest : public BaseTest {
+ protected:
+  TrackingMemoryResource tracking_memory_resource;
+};
+
+TEST_F(TrackingMemoryResourceTest, PerformAllocations) {
+  auto int_ptr = reinterpret_cast<int*>(tracking_memory_resource.allocate(sizeof(int)));
+  EXPECT_NO_THROW((*int_ptr) = 1);
```
[the remark this answers is not in the document]
Comment (twollnik) [reply src/test/lib/memory/tracking_memory_resource_test.cpp:13]: result: it doesnt. my suggestion: remove the test

File: src/lib/memory/tracking_memory_resource.cpp [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+void TrackingMemoryResource::do_deallocate(void* p, std::size_t bytes, std::size_t alignment) {
+  std::free(p);
+  _memory_timeseries.emplace_back(std::make_pair(_get_timestamp(), -1 * bytes));
+}
+
+[[nodiscard]] bool TrackingMemoryResource::do_is_equal(const memory_resource& other) const BOOST_NOEXCEPT {
+  return &other == this;
+}
+
+const std::vector<std::pair<std::chrono::system_clock::time_point, int64_t>>&
```
[the remark this answers is not in the document]
Comment (janlemcke) [reply src/lib/memory/tracking_memory_resource.cpp:23]: Sollten/Können wir das dann auch im Rückgabewert der Funktion `TrackingMemoryResource::_get_timestamp()` nutzen? 

Aber an sich sehr gute Idee.

File: src/lib/memory/tracking_memory_resource.hpp [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+
+namespace opossum {
+
+class TrackingMemoryResource : public boost::container::pmr::memory_resource {
+ public:
+  TrackingMemoryResource() = default;
+  void* do_allocate(std::size_t bytes, std::size_t alignment) override;
+  void do_deallocate(void* p, std::size_t bytes, std::size_t alignment) override;
+  [[nodiscard]] bool do_is_equal(const memory_resource& other) const BOOST_NOEXCEPT override;
+  const std::vector<std::pair<std::chrono::system_clock::time_point, int64_t>>& memory_timeseries() const;
```
Comment (janlemcke) [inline src/lib/memory/tracking_memory_resource.hpp:17]: Ebenfalls anpassen für:
`using timestamp = std::chrono::system_clock::time_point;`

File: src/lib/operators/join_hash.cpp [hunk 12]
```diff
@@ -9,11 +9,9 @@
 #include <utility>
 #include <vector>
 
-#include "bytell_hash_map.hpp"
```
Comment (janlemcke) [inline src/lib/operators/join_hash.cpp:12]: Everthing regarding changing the map should be described in the PR + results from benchmarks?

File: src/test/lib/utils/memory_resource_manager_test.cpp [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+  // Will be executed by multiple threads simultaneously.
+  auto fetch_and_use_a_memory_resource = [&](const uint8_t indx) {
+    auto memory_resource =
+        memory_resource_manager.get_memory_resource(OperatorType::Mock, "my_data_structure_" + std::to_string(indx));
+    auto mem_ptr = memory_resource->allocate(indx + 1);
+    memory_resource->deallocate(mem_ptr, indx + 1);
+  };
+
+  // Create a few threads that interact with the memory resource manager and obtained memory resources
+  auto N_THREADS = uint8_t{10};
```
Comment (tjjordan) [inline src/test/lib/utils/memory_resource_manager_test.cpp:59]: const

File: src/test/lib/utils/memory_resource_manager_test.cpp [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+  // Ensure that the number of memory resources matches the number of threads.
+  const auto memory_resources = memory_resource_manager.memory_resources();
+  ASSERT_EQ(memory_resources.size(), N_THREADS);
+
+  // The total allocated amount should be as expected.
+  // We expect a total of 1+2+..+N_THREADS = ((N_THREADS^2 + N_THREADS) / 2)
+  // bytes to have been allocated or deallocated.
+  auto n_allocated_bytes = int{0};
+  auto n_deallocated_bytes = int{0};
+  auto expected_allocation_amount = (N_THREADS * N_THREADS + N_THREADS) / 2;
```
Comment (tjjordan) [inline src/test/lib/utils/memory_resource_manager_test.cpp:77]: const

File: src/lib/operators/join_hash.cpp [hunk 15]
```diff
@@ -383,7 +387,7 @@ class JoinHash::JoinHashImpl : public AbstractReadOnlyOperatorImpl {
      */
     if (_radix_bits > 0) {
       Timer timer_clustering;
-      auto jobs = std::vector<std::shared_ptr<AbstractTask>>{};
+      std::vector<std::shared_ptr<AbstractTask>> jobs;
```
Comment (tjjordan) [inline src/lib/operators/join_hash.cpp:390]: Why was this change made?

File: src/lib/operators/join_hash/join_hash_steps.hpp [hunk 16]
```diff
@@ -60,7 +66,7 @@ struct Partition {
   // Bit vector to store NULL flags - not using uninitialized_vector because it is not specialized for bool.
   // It is stored independently of the elements as adding a single bit to PartitionedElement would cause memory waste
   // due to padding.
-  std::vector<bool> null_values;
+  pmr_vector<bool> null_values = pmr_vector<bool>(alloc<bool>("69 Steps | Partition struct"));
```
Comment (twollnik) [inline src/lib/operators/join_hash/join_hash_steps.hpp:69]:
```suggestion
  auto null_values = pmr_vector<bool>(alloc<bool>("null_flags_bitvector_partition"));
```

**Pull request merged**